### PR TITLE
ci: make LLxprt PR reviewer non-blocking (Fixes #2778)

### LINKED ISSUE #2778: Make the informational LLxprt PR reviewer non-blocking
## Problem

The LLxprt walkthrough is informational, but `.github/workflows/pr-review.yml` currently lets infrastructure and reviewer-service failures conclude the `Run LLxprt review` job as failed. On PR 2750, both attempts successfully wrote and posted the generic fallback comment, yet the nonzero walkthrough process left the PR with a failed check and `UNSTABLE` merge state.

An informational reviewer must preserve truthful internal failures and diagnostics without becoming a merge gate.

## Evidence

- PR 2750 run 30271261051 failed twice with `Walkthrough pipeline failed: Cannot parse JSON from response`.
- Installation, quota selection, fallback creation, diagnostic upload, and comment posting completed; no source finding was emitted.
- The workflow's review job has no final non-blocking boundary.
- The walkthrough step propagates its process exit status.
- Diagnostic upload runs only on `failure()`, so parse artifacts produced by graceful degradation on a zero-exit pipeline are not uploaded.
- Fallback creation and posting use `always()`, but their failures can still affect the job.

## Required semantics

Internal commands should remain fail-fast and return nonzero when they fail. The GitHub Actions workflow boundary should contain those outcomes because this reviewer is advisory.

## Bounded plan

1. Add a final job-level non-blocking boundary for the review job.
2. Capture install and quota failures as explicit step outcomes and logs without ending finalization.
3. Skip model invocation when install or quota prerequisites fail, avoiding misleading secondary errors.
4. Run the walkthrough with a contained step outcome while preserving its real exit status and stderr diagnostics.
5. Upload available install, quota, walkthrough, and raw parse diagnostics under `always()`; artifact upload itself must be non-blocking.
6. Always attempt fallback creation and tagged comment upsert; both finalization steps must be non-blocking.
7. Add workflow-structure tests and injected-failure behavior coverage for every boundary.

## Acceptance criteria

For injected failures in nightly installation, quota/key selection, model invocation, JSON parsing, diagnostic upload, and comment posting:

- [ ] The LLxprt reviewer does not conclude as a blocking failure or destabilize the PR.
- [ ] The original failing command's nonzero outcome remains visible in step outcome and logs.
- [ ] Installation or quota failure skips model invocation.
- [ ] A generic fallback comment is generated whenever no substantive walkthrough exists.
- [ ] Tagged comment posting is attempted after earlier failures.
- [ ] Comment-posting failure is non-blocking.
- [ ] Available installation, quota, walkthrough, and raw parse diagnostics are uploaded with the existing bounded retention.
- [ ] Parse artifacts from gracefully degraded zero-exit runs are uploaded.
- [ ] Raw responses, secrets, and internal diagnostics are never copied into public comments.
- [ ] Linked-issue gating and mergeability-gate behavior remain unchanged.
- [ ] Cancellation remains cancellation; the workflow does not attempt to disguise explicit cancellation as review success.
- [ ] Workflow tests verify exact conditions, ordering, pinned actions, and comment tag behavior.
- [ ] Test, lint, typecheck, format, and build pass without suppressions or weakened rules.

## Non-goals

- Making model or parser failures look like successful reviews inside logs.
- Swallowing errors inside reviewer scripts.
- Changing linked-issue requirements or the separate mergeability gate.
- Changing OCR, CodeRabbit, main CI, branch protection, or repository-wide required-check policy.
- Adding retries beyond the existing bounded parsing/network policy.

## Expected paths

- `.github/workflows/pr-review.yml`
- `scripts/tests/pr-review-walkthrough-workflow.test.js`
- A focused workflow behavior fixture/test only if required by the existing test architecture

### PR DESCRIPTION
**Pull request by acoliver**
## Summary

The LLxprt walkthrough reviewer is advisory, but `.github/workflows/pr-review.yml` let infrastructure and reviewer-service failures conclude the `Run LLxprt review` job as failed. On PR 2750, both attempts successfully wrote and posted the generic fallback comment, yet the nonzero walkthrough process left the PR with a failed check and `UNSTABLE` merge state.

This PR adds a job-level non-blocking boundary so internal command failures remain visible in step outcomes and logs without becoming a merge gate.

### `.github/workflows/pr-review.yml`
- **`continue-on-error: true`** on all advisory steps: install, quota, walkthrough, upload, fallback, and comment-post. This preserves each step's real `outcome` (visible in the Actions UI and logs) while preventing any single advisory failure from failing the job.
- **Gate walkthrough on install/quota success**: the `Run walkthrough pipeline` step now has `if: ${{ steps.issue_gate.outputs.should_review == 'true' && steps.install.outcome == 'success' && steps.quota.outcome == 'success' }}`. When install or quota fails, model invocation is skipped entirely, avoiding misleading secondary errors.
- **Change diagnostics upload from `failure()` to `always()`**: a gracefully degraded pipeline may exit 0 while producing parse-failure artifacts. Running the upload under `always()` ensures those artifacts are captured alongside failure logs.
- **IDs added**: `install`, `walkthrough` step IDs for outcome gating.

### `scripts/tests/pr-review-walkthrough-workflow.test.js`
- Updated the existing upload-condition test from `failure()` to `always()`.
- Added 29 new tests covering:
  - Every advisory step has `continue-on-error: true`
  - Walkthrough is skipped when install or quota fails
  - Walkthrough runs when both install and quota succeed
  - Diagnostics upload runs under `always()` for both success and failure
  - Fallback and comment-post always attempted and non-blocking
  - Cancellation is not disguised as success (no fabricated success expressions)
  - Issue gate and mergeability gate remain unchanged
  - Injected-failure scenarios for install, quota, walkthrough, JSON parse, upload, and comment-post failures
  - Secrets/diagnostics never leaked into public comments
- Added a `evalStepIf` helper that evaluates GitHub Actions `if:` expressions against a map of step outcomes, with strict trailing-input validation and correct `success()`/`failure()`/`cancelled()` semantics.

## Acceptance criteria

- [x] The LLxprt reviewer does not conclude as a blocking failure or destabilize the PR.
- [x] The original failing command's nonzero outcome remains visible in step outcome and logs.
- [x] Installation or quota failure skips model invocation.
- [x] A generic fallback comment is generated whenever no substantive walkthrough exists.
- [x] Tagged comment posting is attempted after earlier failures.
- [x] Comment-posting failure is non-blocking.
- [x] Available installation, quota, walkthrough, and raw parse diagnostics are uploaded with the existing bounded retention.
- [x] Parse artifacts from gracefully degraded zero-exit runs are uploaded.
- [x] Raw responses, secrets, and internal diagnostics are never copied into public comments.
- [x] Linked-issue gating and mergeability-gate behavior remain unchanged.
- [x] Cancellation remains cancellation; the workflow does not attempt to disguise explicit cancellation as review success.
- [x] Workflow tests verify exact conditions, ordering, pinned actions, and comment tag behavior.

## Non-goals (unchanged)

- No script changes to `pr-review-walkthrough.mjs` or `ci-quota-check.js`
- No changes to linked-issue requirements or the separate mergeability gate
- No changes to OCR, CodeRabbit, main CI, branch protection, or repository-wide required-check policy
- No retries beyond the existing bounded parsing/network policy

## Test plan

- [x] `npm run test:scripts` — 76 workflow tests pass
- [x] `npm run lint` — clean (no suppressions or weakened rules)
- [x] `npm run format` — clean
- [x] `npm run build` — passes
- [x] Open Code Review (ocr) — 3 findings addressed (trailing-input validation, success() semantics, output comparison modeling)

Closes #2778

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved automated review reliability when installation or quota checks fail.
  * Review walkthroughs now run only when required setup steps succeed.
  * Diagnostics are uploaded even when walkthrough processing fails.
  * Commenting and fallback actions no longer block the overall review process.
  * Preserved generic fallback comments and prevented diagnostic output from appearing in review comments.

### CODE REVIEW

File: scripts/tests/pr-review-walkthrough-workflow.test.js [hunk 1]
```diff
@@ -36,6 +36,133 @@ function allStepText(job) {
...
+      }
+      return value;
+    }
+    if (expr[pos] === '!') {
+      pos += 1;
+      skipSpaces();
+      return !parsePrimary();
+    }
+    return false;
+  }
```
[the remark this answers is not in the document]
Comment (acoliver) [reply scripts/tests/pr-review-walkthrough-workflow.test.js]: Valid finding. Fixed in 47d783d0e: `parsePrimary()` now throws on unexpected tokens or end-of-input (including missing closing parens), so malformed expressions like `TRUE &&`, `TRUE ||`, or bare `!` are caught early instead of silently returning false. The trailing-input check at the end of `evalBoolean` still guards against any leftover tokens. All 76 tests pass with the stricter parser.

File: scripts/tests/pr-review-walkthrough-workflow.test.ts [hunk 2]
```diff
@@ -0,0 +1,835 @@
...
+  normalized = normalized.replace(
+    /steps\.([a-zA-Z0-9_-]+)\.outputs\.([a-zA-Z0-9_-]+)\s*==\s*'([^']*)'/g,
+    (_match, stepId: string, outputName: string, expected: string) => {
+      const actual = outputs[stepId]?.[outputName];
+      if (actual === undefined) {
+        return 'TRUE';
+      }
+      return actual === expected ? 'TRUE' : 'FALSE';
+    },
+  );
```
[the remark this answers is not in the document]
Comment (acoliver) [reply scripts/tests/pr-review-walkthrough-workflow.test.ts]: Addressed in the latest push. The `requireStepById` helper is now extracted as a single top-level function that asserts step existence and returns `WorkflowStep`. Both the non-blocking and injected-failure describe blocks now use it (the non-blocking block via a thin `stepById` wrapper, the injected-failure block via direct `find` for the optional-undefined pattern). The inline comment on the default-to-TRUE behavior now documents that critical outputs (should_review, selected_key) are always provided by tests that exercise those paths, so tests cannot silently pass when a critical output is missing.

File: scripts/tests/pr-review-walkthrough-workflow.test.ts [hunk 3]
```diff
@@ -0,0 +1,835 @@
...
+
+  describe('non-blocking review job (Issue #2778)', () => {
+    function stepById(id: string): WorkflowStep {
+      const step = reviewJob.steps.find((s) => s.id === id);
+      expect(
+        step,
+        `review job should have a step with id "${id}"`,
+      ).toBeTruthy();
+      return step as WorkflowStep;
+    }
```
[the remark this answers is not in the document]
Comment (acoliver) [reply scripts/tests/pr-review-walkthrough-workflow.test.ts:539]: Addressed in the latest push. Extracted a single top-level `requireStepById(job, id)` helper that asserts existence and returns `WorkflowStep`. The non-blocking describe block uses it via a thin `stepById` wrapper. The injected-failure block intentionally uses `find` returning `WorkflowStep | undefined` because those tests pass the result to `continueOnError()` which handles `undefined` by returning `false`. The two patterns now have clearly distinct semantics rather than being silent duplications.

File: scripts/tests/pr-review-walkthrough-workflow.test.ts [hunk 4]
```diff
@@ -0,0 +1,836 @@
...
+
+    function stepByName(name: string): WorkflowStep {
+      const step = reviewJob.steps.find((s) => s.name === name);
+      expect(
+        step,
+        `review job should have a step named "${name}"`,
+      ).toBeTruthy();
+      return step as WorkflowStep;
+    }
```
[the remark this answers is not in the document]
Comment (acoliver) [reply scripts/tests/pr-review-walkthrough-workflow.test.ts:543]: Addressed in commit 7108174bd. Extracted a single top-level `requireStepByName` helper to module scope, eliminating the duplicate `stepByName` in the non-blocking describe block. The injected-failure block intentionally uses optional `find()` returning `WorkflowStep | undefined` because those tests pass results to `continueOnError()` which handles undefined by returning false. This semantic difference is now documented with a comment explaining why the two patterns coexist.

File: scripts/tests/pr-review-walkthrough-workflow.test.ts [hunk 5]
```diff
@@ -0,0 +1,836 @@
...
+
+    // --- A2/A3: Install/quota failure is visible and skips walkthrough ---
+
+    it('walkthrough step is skipped when install fails', () => {
+      const walkthrough = stepById('walkthrough');
+      expect(
+        evalStepIf(walkthrough.if, {
+          outcomes: { install: 'failure', quota: 'success' },
+        }),
+      ).toBe(false);
```
[the remark this answers is not in the document]
Comment (acoliver) [reply scripts/tests/pr-review-walkthrough-workflow.test.ts:591]: Addressed in commit 7108174bd. All walkthrough step tests now provide `outputs: { issue_gate: { should_review: 'true' } }` so the full gating condition (`steps.issue_gate.outputs.should_review == 'true' && steps.install.outcome == 'success' && steps.quota.outcome == 'success'`) is exercised. Added a new test verifying the walkthrough does not run when `should_review` is 'false'.

File: scripts/tests/pr-review-walkthrough-workflow.test.ts [hunk 6]
```diff
@@ -0,0 +1,836 @@
...
+
+function evalBoolean(expr: string): boolean {
+  let pos = 0;
+  function skipSpaces(): void {
+    while (pos < expr.length && expr[pos] === ' ') {
+      pos += 1;
+    }
+  }
+  function parsePrimary(): boolean {
+    skipSpaces();
```
[the remark this answers is not in the document]
Comment (acoliver) [reply scripts/tests/pr-review-walkthrough-workflow.test.ts:169]: Addressed in commit 7108174bd. Added an early return `true` for empty or whitespace-only expressions in `evalBoolean`, matching GitHub Actions semantics where an empty `if:` is always-true.

File: scripts/tests/pr-review-walkthrough-workflow.test.ts [hunk 7]
```diff
@@ -0,0 +1,836 @@
...
+  });
+
+  describe('injected-failure scenario coverage (Issue #2778)', () => {
+    function stepByName(name: string): WorkflowStep | undefined {
+      return reviewJob.steps.find((s) => s.name === name);
+    }
+
+    function stepById(id: string): WorkflowStep | undefined {
+      return reviewJob.steps.find((s) => s.id === id);
+    }
```
[the remark this answers is not in the document]
Comment (acoliver) [reply scripts/tests/pr-review-walkthrough-workflow.test.ts:728]: Addressed in commit 7108174bd. The non-blocking block now delegates to the shared top-level `requireStepByName` helper, eliminating the duplication. The injected-failure block's optional `find()` pattern is intentionally retained (see reply on the companion comment) and documented.
